### DISCUSSION TITLE
named url parameter support in `request-data` processor

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#720] [General]
<details><summary>Status: In Progress</summary>

Follow: https://github.com/ftd-lang/fpm/discussions/526
Code PR: https://github.com/ftd-lang/fpm/pull/566
Doc PR: https://github.com/ftd-lang/fpm.dev/pull/89

</details>

For this feature, the URL itself would be dynamic, eg `/amitu/`. Question is how would FPM know which ftd file to use if request comes to `/<username>/`?

For this to work we need [dynamic url in sitemap feature](https://github.com/ftd-lang/fpm/discussions/311) support.

```ftd
-- string username:
$processor$: single-request-data
```

Should just work. The variable name is `username`, so we will look for `username` in the dynamic URLs. `named > query > post`.

This will also work:

```ftd
-- record request-data:
username string:

-- request-data rd:
$procesor$: request-data

-- $rd.username
```

What if post body was:

```json
{
    "rd": { "username": "foo" },
    "username": "bar"
}
```

What would be `$rd.username`? With `single-request-data` it would be `foo` and with `request-data` it would be `bar`.

**Comment by AbrarNitk** [comment]
## Implementation Approach

When a request comes for `/abrark/foo/28/`, we are parsing and matching it with sitemap `url: /<string:username>/foo/<integer:age>/`, while matching we will store path parameters `[(username, abrark), (age, 28)]` to `config.request.path_params`. Now to get this information in `request-data` `processor` Config will provide a `function` say `config.request.get_path_params` which will return the `config.request.path_params` value.

**Reply by AbrarNitk** [reply]
Question: How do store generic information in config.request.path_params, what should be the type of it?

**Reply by AbrarNitk** [reply]
The answer to the above question is `ftd::Value`.
